### ISSUE TITLE
replication failure with type=class

### ISSUE THREAD

**Issue by markeschaffer** [aahrens1/pystacked#16]
See example below.

- For type=reg with the norandom option and reg+lassocv, changing the order of the two learners doesn't change the stacking weights.
- For type=class with the norandom option and logit+lassocv, changing the order of the two learners leads to different weights.

This is a general problem with type=class but is the easiest way to illustrate it.  Code below is in the Dropbox folder.

```
cap prog drop pystacked
cap prod drop pystacked_p

// auto data
sysuse auto, clear
drop if rep78==.
sort weight

set seed 123
// classification: change order, replication fails, weights different
pystacked foreign price-gear if _n<=60, type(class) methods(logit lassocv) folds(3) norandom
pystacked foreign price-gear if _n<=60, type(class) methods(lassocv logit) folds(3) norandom

set seed 123
// regression: change order, replication works, weights identical
pystacked price mpg-gear if _n<=60, type(reg) methods(ols lassocv) folds(3) norandom
pystacked price mpg-gear if _n<=60, type(reg) methods(lassocv ols) folds(3) norandom
```

**Comment by aahrens1** [comment]
Works fine if you set the seed again: 

```
cap prog drop pystacked
cap prod drop pystacked_p

// auto data
sysuse auto, clear
drop if rep78==.
sort weight

set seed 123
// classification: change order, replication fails, weights different
pystacked foreign price-gear if _n<=60, type(class) methods(logit lassocv) folds(3) norandom
set seed 123
pystacked foreign price-gear if _n<=60, type(class) methods(lassocv logit) folds(3) norandom

set seed 123
// regression: change order, replication works, weights identical
pystacked price mpg-gear if _n<=60, type(reg) methods(ols lassocv) folds(3) norandom
set seed 123
pystacked price mpg-gear if _n<=60, type(reg) methods(lassocv ols) folds(3) norandom
```

**Comment by markeschaffer** [comment]
But setting the seed shouldn't be necessary at all with the norandom option and these learners. That's the behaviour with type=reg. I suspect norandom is being ignored if type=class.

**Comment by aahrens1** [comment]
> I suspect norandom is being ignored if type=class.
I double checked. "nonrandom" creates a fold variable in Stata that is passed on to Python's stacking regressor and classifier. All good. 

The second `pystacked' call starts from a different Stata & Python seed. The seed is still important for cross-validating the lasso. That's why you should reset the seed. 

The actual puzzle is why stacking regression still yields the same results without resetting the seed.

**Comment by markeschaffer** [comment]
This is interesting. I thought from the behaviour of pystacked with lassocv+norandom that norandom also applied to lassocv, but apparently not. (Maybe it should?)

Maybe the default behaviour of linear lassocv is different from the default behaviour of logit lassocv.

**Comment by markeschaffer** [comment]
I've checked the behaviour of sklearn's LassoCV and LogisticRegressionCV. Defaults are not fully spelled out, unfortunately.

In principle, one should be able to pass them an option like

`cv(sklearn.model_selection.KFold(n_splits=5,shuffle=False))`

but pystack's parser allows only `cv(int)`.

I will try to check in python on this computer but my guess is that LassoCV defaults to shuffle=False and LogisticRegressionCV defaults to shuffle=True.

**Comment by markeschaffer** [comment]
I was close but not quite right - the default shuffling is done by the solver, not by the CV splitter. If random_state is provided as an argument, it replicates.

BTW, there's a tiny bug in _pyparse that I've fixed. random_state=0 is a legit setting that is used in some examples on the sklearn website, but _pyparse said `if `random_state'>0` (and I couldn't replicate at first). Have changed it to `if `random_state'>=0`.

**Comment by aahrens1** [comment]
We should probably change the default to shuffle=True. I find it odd that CV splits aren't randomized. What do you think?

**Comment by markeschaffer** [comment]
Maybe ... not sure. It's handy for replication if the splits use the current order.  Could perhaps make the norandom option apply to the CV estimators as well?

**Comment by chansen776** [comment]
@aahrens1 @markeschaffer I also find it weird that CV splits aren't randomized. It's not a big deal at some level, but it's a bit weird as default behavior to me.

**Comment by aahrens1** [comment]
Replication works either way. We only need to set the seed at the beginning of the script. 

norandom is essentially a developer option, mostly for us. 

I think people expect random fold split when CV is done.

**Comment by aahrens1** [comment]
Thanks @chansen776, agree. 

Again one of those weird scikit-learn defaults.

**Comment by markeschaffer** [comment]
Replication is (much) easier if there's direct control over the folds - that's why the norandom option was helpful for me when tracking this down.  It's not just replication of pystacked with pystacked.  I'm also using the new Stata stacking command to replicate pystacked.  That's how I found out what was going on in this example.

The problem is that pystacked currently doesn't allow the cv option to accept a CV splitter. That would allow something like

`cv(sklearn.model_selection.KFold(n_splits=5,shuffle=False))`

as an option for a specified learner.

I'd be happy to make shuffle=True as the default for a learner that uses a CV splitter if we can also turn it off.

**Comment by aahrens1** [comment]
> The problem is that pystacked currently doesn't allow the cv option to accept a CV splitter. 

I'll add that functionality.

**Comment by markeschaffer** [comment]
And/or a simple "noshuffle" Stata option for the individual learners.

**Comment by aahrens1** [comment]
I have added the `noshuffle` option. The default is that cross-validated base learners randomize the cv fold split.
